File: examples/java/src/main/java/runar/examples/p256-primitives/P256Primitives.runar.java
package runar.examples.p256primitives;

import java.math.BigInteger;

import runar.lang.SmartContract;
import runar.lang.annotations.Public;
import runar.lang.annotations.Readonly;
import runar.lang.types.P256Point;

import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.p256Add;
import static runar.lang.Builtins.p256Mul;
import static runar.lang.Builtins.p256MulGen;
import static runar.lang.Builtins.p256OnCurve;

/**
 * P256Primitives -- conformance fixture for the NIST P-256 (secp256r1)
 * EC primitives. Each method recomputes the stored {@code expectedPoint}
 * via a different primitive ({@code p256Mul}, {@code p256Add},
 * {@code p256MulGen}) and asserts the result equals the stored value.
 */
class P256Primitives extends SmartContract {

    @Readonly P256Point expectedPoint;

    P256Primitives(P256Point expectedPoint) {
        super(expectedPoint);
        this.expectedPoint = expectedPoint;
    }

    @Public
    void verify(BigInteger k, P256Point basePoint) {
        P256Point result = p256Mul(basePoint, k);
        assertThat(p256OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }

    @Public
    void verifyAdd(P256Point a, P256Point b) {
        P256Point result = p256Add(a, b);
        assertThat(p256OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }

    @Public
    void verifyMulGen(BigInteger k) {
        P256Point result = p256MulGen(k);
        assertThat(p256OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }
}

File: examples/go/p256-primitives/P256Primitives.runar.go
//go:build ignore

// EXCLUDED FROM THE GO BUILD — a P-256 scalar does not fit the Go mock's
// `Bigint`.
//
//	cannot use k (variable of int64 type runar.Bigint) as *big.Int value in
//	argument to runar.P256Mul
//
// `Verify(k runar.Bigint, …)` is correct Rúnar: the parser maps `Bigint` to the
// arbitrary-precision `bigint`, and a P-256 scalar is 256 bits. packages/runar-go
// is also correct: P256Mul takes a `*big.Int` because that is the only Go type
// that can hold one. The two cannot meet while `runar.Bigint` aliases int64.
//
// Widening the mock's `Bigint` is the fix, and it is a change to every tier's
// notion of the primitive plus roughly forty call sites across examples/go —
// not something to smuggle in behind a build tag. Same root cause as
// schnorr-zkp, integer-boundary and go-dsl-bytestring-literal, and the same one
// behind the `knownDivergent` EcPointX/EcPointY entries in
// packages/runar-go/mock_script_agreement_test.go.

package contract

import "runar"

type P256Primitives struct {
	runar.SmartContract
	ExpectedPoint runar.P256Point `runar:"readonly"`
}

func (c *P256Primitives) Verify(k runar.Bigint, basePoint runar.P256Point) {
	result := runar.P256Mul(basePoint, k)
	runar.Assert(runar.P256OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}

func (c *P256Primitives) VerifyAdd(a runar.P256Point, b runar.P256Point) {
	result := runar.P256Add(a, b)
	runar.Assert(runar.P256OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}

func (c *P256Primitives) VerifyMulGen(k runar.Bigint) {
	result := runar.P256MulGen(k)
	runar.Assert(runar.P256OnCurve(result))
	runar.Assert(result == c.ExpectedPoint)
}
